src/com/linkedin/parseq/Engine.java

```diff
@@ -103,6 +96,37 @@
     _allLogger = loggerFactory.getLogger(LOGGER_BASE + ":all");
     _rootLogger = loggerFactory.getLogger(LOGGER_BASE + ":root");
 
+    if (_properties.containsKey(MAX_RELATIONSHIPS_PER_TRACE)) {
+      _maxRelationshipsPerTrace = (Integer) getProperty(MAX_RELATIONSHIPS_PER_TRACE);
+    } else {
+      _maxRelationshipsPerTrace = DEFUALT_MAX_RELATIONSHIPS_PER_TRACE;
+    }
+
+    if (_properties.containsKey(MAX_CONCURRENT_PLANS)) {
+      _maxConcurrentPlans = (Integer) getProperty(MAX_CONCURRENT_PLANS);
+    } else {
+      _maxConcurrentPlans = DEFUALT_MAX_CONCURRENT_PLANS;
+    }
+    _concurrentPlans = new Semaphore(_maxConcurrentPlans);
```

Comment:
Do we want FIFO guarantee on the semaphore? It seems counter intuitive to me that the plan submitted first may fail with a timeout but the plan submitted later can go through.